ENH/API: ExtensionArray.factorize

TomAugspurger opened the pull request:
Adds factorize to the interface for ExtensionArray, with a default
implementation.

This is a stepping stone to groupby.

Categorical already has a custom implementation.

Decimal reuses the default.

JSON can't use the default since an array of dictionaries isn't hashable, so we use a custom implementation.

TomAugspurger committed:
ENH: Sorting of ExtensionArrays

This enables {Series,DataFrame}.sort_values and {Series,DataFrame}.argsort

TomAugspurger committed:
REF: Split argsort into two parts

TomAugspurger committed:
Fixed docstring

TomAugspurger committed:
Remove _values_for_argsort

TomAugspurger committed:
Revert "Remove _values_for_argsort"

This reverts commit 44b6d72be5b2431bbe38454a7fc7439244575925.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+sort

TomAugspurger committed:
Workaround Py2

TomAugspurger committed:
Indexer as array

TomAugspurger committed:
Fixed dtypes

TomAugspurger committed:
Fixed docstring

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+sort

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+sort

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+sort

TomAugspurger committed:
ENH/API: ExtensionArray.factorize

Adds factorize to the interface for ExtensionArray, with a default
implementation.

This is a stepping stone to groupby.

TomAugspurger committed:
Update docs

TomAugspurger commented on pandas/tests/extension/json/array.py:
```diff
@@ -104,6 +105,21 @@ def _concat_same_type(cls, to_concat):
+        frozen = tuple(tuple(x.items()) for x in self)
+        labels, uniques = pd.factorize(frozen)
+
+        # fixup NA
```
This type of thing is going to be error prone. We expect that `[B, NA, C, B]` is coded as `[0, -1, 1, 0]`, not `[0, 1, 2, 0]` (I ran into this with cyberpandas as well).

If library authors are using our tests, this should be caught. Otherwise, I'm not sure how to handle it. We could design a method like around `def refactorize(labels, mask, uniques)` that does something like this, but I'm not sure.

TomAugspurger replied:
Hmm apparently I messed this up, so the tests will fail. Fixing now.

TomAugspurger replied:
Fixed in 434df7dc9c927ac52b881e748ebee12c1389cf15

jorisvandenbossche replied:
If we require this, I think we should put a stronger comment (maybe as a comment below the docstring) about that we expect that missing values are not *in* the uniques.

What would go wrong if `.factorize()` on an extension array would include its missing value in the uniques?

TomAugspurger replied:
Hmm probably not the end of the world, but it would violate our factorize (and eventually groupby) semantics. Until recently, we messed this up for Categorical. That didn't have downstream effects for groupby, since everything in groupby is special for categorical.

I'll make this clearer.

TomAugspurger replied:
Well adding that simple note turned into a rabit hole. I decided to merge the 3 docstrings

- pandas.factorize
- pandas.core.base.IndexOpsMixin.factorize
- pandas.Categorical.factorize

TomAugspurger committed:
fixup! ENH/API: ExtensionArray.factorize

TomAugspurger commented:
Marking this as a WIP for now. For JSONArray, the `value_counts` test is failing. I'd like to see if we can reuse `factorize` for `value_counts`. In principal, a counting the labels and doing a `uniques.take` on the index of the counts should work.

TomAugspurger committed:
REF: Changed ExtensionDtype inheritance

`is_extension_array_dtype(dtype)` was incorrect for dtypes that haven't
implemented the new interface yet. This is because they indirectly subclassed
ExtensionDtype.

This PR changes the hierarchy so that PandasExtensionDtype doesn't subclass
ExtensionDtype. As we implement the interface, like Categorical, we'll add
ExtensionDtype as a base class.

Before:

```
DatetimeTZDtype <- PandasExtensionDtype <- ExtensionDtype (wrong)
CategoricalDtype <- PandasExtensionDtype <- ExtensionDtype (right)

After:

DatetimeTZDtype <- PandasExtensionDtype
                                        \
                                         - _DtypeOpsMixin
                                        /
                   ExtensionDtype ------

CategoricalDtype - PandasExtensionDtype -
                \                        \
                 \                        -_DtypeOpsMixin
                  \                      /
                   ExtensionDtype -------

```

Once all our extension dtypes have implemented the interface we can go back
to the simple, linear inheritance structure.

TomAugspurger commented:
The failures are caused by a "bug" in is_extension_dtype. https://github.com/pandas-dev/pandas/pull/20363

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger commented on pandas/core/algorithms.py:
```diff
@@ -502,9 +501,9 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
+    if is_extension_array_dtype(values):
         values = getattr(values, '_values', values)
-        labels, uniques = values.factorize()
+        labels, uniques = values.factorize(na_sentinel=na_sentinel)
```
Note: this was a bug in #19938 where I forgot to pass this through. It's covered by our extension tests.

TomAugspurger commented:
This should pass now.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -248,6 +248,41 @@ def unique(self):
+        -----
+        :meth:`pandas.factorize` offers a `sort` keyword as well.
+        """
+        from pandas.core.algorithms import _factorize_array
```
We're OK with using this private API here? 

Because an extension authors might want to copy paste this method and change the `arr = self.astype(object)` line? (how do you implement `factorize` in cyberpandas?)

TomAugspurger replied:
Quite similar to this. https://github.com/ContinuumIO/cyberpandas/blob/468644bcbdc9320a1a33b0df393d4fa4bef57dd7/cyberpandas/base.py#L72

In that case I think going to object dtypes is unavoidable, since there's no easy way to factorize a 2-D array, and I didn't want to write a new hashtable implementation :)

TomAugspurger replied:
w.r.t. using factorize array, I don't think it's avoidable.

We might consider making it public / semi-public (keep the `_` to not confuse users, but point EA authors to it).

jorisvandenbossche commented on pandas/tests/extension/base/methods.py:
```diff
@@ -42,3 +43,22 @@ def test_unique(self, data, box, method):
+
+    def test_factorize_equivalence(self, data_for_grouping):
+        l1, u1 = pd.factorize(data_for_grouping)
+        l2, u2 = pd.factorize(data_for_grouping)
```
What is this test doing?

TomAugspurger replied:
Ha, I think I meant for one to be `data_for_grouping.factorize()`

jorisvandenbossche commented on pandas/util/testing.py:
```diff
@@ -1083,6 +1084,32 @@ def _raise(left, right, err_msg):
+    Missing values are checked separately from valid values.
+    A mask of missing values is computed for each and checked to match.
+    The remaining all-valid values are cast to object dtype and checked.
+    """
```
Why was this needed only now? (wasn't the missing values the reason you added `assert_frame_equal` et al as overridable class methods?)

TomAugspurger replied:
`factorize` is I think the first case where we have a public method returning an ExtensionArray (`unique`).

I'll see if any old tests can make use of this.

TomAugspurger replied:
I don't see any other cases where we could apply `assert_extension_array_equal`.

The closes would be the `__getitem__` tests, but in that case we don't have an expected array to compare to, since we're testing getitem / take / etc.

TomAugspurger committed:
Fix factorize equivalence test

TomAugspurger committed:
Shared factorize doc

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -248,6 +248,48 @@ def unique(self):
         return self._constructor_from_sequence(uniques)
 
+    def factorize(self, na_sentinel=-1):
+        """Encode the extension array as an enumerated type.
```
Ah, need to merge this one too maybe. Will have to check on import order...

TomAugspurger commented:
Sorry about the tangentially related changes refactoring the docstring. I can revert if needed, but I don't think it complicates review too much since the only non-doc change to `algos.py` is replacing an `is_categorical` with an `is_extension_array_dtype`.

Tracking down a py-2 failure now.

TomAugspurger committed:
Move to algorithms

TomAugspurger committed:
BUG: py2 bug

TomAugspurger commented:
Here's the new docstring for pandas.factorize

![fireshot capture 005 - pandas factorize pandas 0 23 0 dev0_ - file____users_taugspurger_sandbox_](https://user-images.githubusercontent.com/1312546/37541452-ad85b388-2928-11e8-8b71-e300ff9544af.png)

TomAugspurger committed:
Typo, ref

jreback commented:
hmm, ``order`` is still there I thought we removed that? (or maybe we just deprecated it, can you add back as a deprecated parm)?

also, as a public interface, may not really need ``size_hint`` (as now have ``_factorize_array`` which can be an internal interface, but this is small potatoes).

jreback commented:
maybe link to ``pd.unique`` as well

jreback commented on pandas/core/algorithms.py:
```diff
@@ -464,32 +468,117 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
+
+    See Also
+    --------
+    pandas.cut : Discretize continuous-valued array.
```
add ``pandas.unique`` here

TomAugspurger commented:
> can you add back as a deprecated parm

yeah, lost that in the shuffle.

> also, as a public interface, may not really need size_hint

Agreed.

> maybe link to pd.unique as well

👍

jreback commented on pandas/core/algorithms.py:
```diff
@@ -464,32 +468,117 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
+    %(values)s%(sort)s
     na_sentinel : int, default -1
-        Value to mark "not found"
-    size_hint : hint to the hashtable sizer
```
need to show ``order`` as deprecated

jreback commented:
needs a rebase

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -463,32 +468,117 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
-def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
-    """
-    Encode input values as an enumerated type or categorical variable
+_shared_docs['factorize'] = """
```
If it is only a single entry, can we make this a variable (eg `_shared_docstring_factorize`) ?

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -463,32 +468,117 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
+    >>> labels, uniques = pd.factorize(cat)
+    >>> labels
+    array([0, 0, 1])
+
```
remove blank line here

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -463,32 +468,117 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
+
+@Substitution(
+    values=dedent("""\
+    values : Sequence
```
I would personally use something else as "Sequence", I don't think we use that somewhere else (or at least lower case "sequence"). It's "array-like" ?

TomAugspurger replied:
Sorry, was thinking in `typing` terms.

I'm trying to use `array-like` to mean sequence + dtype (which is what `is_array_like` checks).

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -463,32 +468,117 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
+        coereced to ndarrays before factorization.
+    """),
+    sort=dedent("""\
+    sort : boolean, default False
```
boolean -> bool

TomAugspurger committed:
Change name

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+sort

TomAugspurger committed:
Fixed docs

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger committed:
Wording

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+sort

TomAugspurger commented:
Do we want to have `ExtensionArray.factorize` use the shared docstring as well? There's an import order issue, I've tried to keep `arrays/base.py` from importing much from pandas (just `AbstractMethodError` so far. If we do a shared doc I would have to move the docstring to a file on it's own, and then import it from `algorithms.py`, `arrays/base.py`, `core/base.py` and `categorical.py`.

I also haven't really thought through if that would break anything for EA implementers.

TomAugspurger committed:
fixup! Wording

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger committed:
Back to _values_for_argsort

TomAugspurger committed:
Example with _from_factorize

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -301,6 +319,60 @@ def unique(self):
+        """
+        return self.astype(object)
+
+    def factorize(self, na_sentinel=-1):
```
you need a sort=False arg here

TomAugspurger replied:
Do we want / need that? It complicates the implementation a bit. Any idea what it's actually used for?

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -301,6 +319,60 @@ def unique(self):
+
+        mask = self.isna()
+        arr = self._values_for_factorize()
+        arr[mask] = np.nan
```
shouldn't this be the ``self._na_value`` ?

TomAugspurger replied:
We removed that from the API. Anyway, that would be the type-specific NA value, we need an NA value that pandas knows about, which is just NaN or NaT.

Speaking of which, I don't really like how this is done right now. If your values_for_factorize are integers, this will force them to be floating. Will thing more about it.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -2043,47 +2044,9 @@ def unique(self):
-        [a, b]
-        Categories (2, object): [a, b]
-        """
         from pandas.core.algorithms import _factorize_array
```
shouldn't the impl change here? (to use the newly created EA facilities)

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger committed:
Unskip most JSON tests

TomAugspurger committed:
Merge branch 'fu1+sort' into ea-factorize-2

TomAugspurger committed:
Overridable na_value too.

This has the EA author specify the NA value before factorization. This satisfies
the Categorical use case.

TomAugspurger commented:
Ugh I mixed up my branches and merged https://github.com/pandas-dev/pandas/pull/19957 into here... Fixing now.

TomAugspurger committed:
Reverted sorting changes

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -301,6 +319,64 @@ def unique(self):
+            An array suitable for factoraization. This should maintain order
+            and be a supported dtype. By default, the extension array is cast
+            to object dtype.
+        na_value : scalar
```
OK, here's the meaningful change from my last round of commits. We have the user specify an ndarray of values to be factorized, and the na value *going into* the factorization routine.

Previously, `Categorical.factorize` had to be implemented, since it used integers and `iNaT` for the array and missing value in `_factorize_array`. So from a code reuse and interface flexibility point of view, this is a good thing.

I dislike how this opens up the internals of pandas' factorization algorithms to the public more. Is it publicly documented anywhere that iNaT is the minimum integer? How would an EA author know to do something like what we do for Categorical (and PeriodArray when we get there)?

TomAugspurger committed:
Remove a bit more argsort

jreback commented:
this looks fine. I am still concerned about the API expansion for implementors here, IOW you have both ``_values_for_factorize`` and now ``_from_factorize``.

jorisvandenbossche commented:
> I dislike how this opens up the internals of pandas' factorization algorithms to the public more. Is it publicly documented anywhere that iNaT is the minimum integer? How would an EA author know to do something like what we do for Categorical (and PeriodArray when we get there)?

But is it needed that the extension author specifies this value? 
Because what are the options? If the array has integer-like dtype, it is iNaT, otherwise it is np.nan ? 
(if that is the case, I think we could deduce that?)


In general, I am wondering if we couldn't make this simpler. If we could somehow extend `factorize` to specify a mask or a na_value, then the `factorize` method on EA could be something like:

```
def factorize(self, na_sentinel=-1):
    mask = self.isna()
    values = ... # this is the part to fill in by the extension array author (equivalent _values_for_factorize)
    labels, uniques = pd.factorize(values, na_sentinel=na_sentinel, mask=mask)
    uniques = ... # coerce back to extension array (_from_factorized)
    return labels, uniques
```

This is of course *very* similar to the current `factorize` implementation, but with the difference that it does not use internal APIs. And actually, overriding this does not take more lines of code than overriding both `_values_for_factorize` and `_from_factorized`. So I am not fully sure of the advantage of providing both ways to do it.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger commented:
> So I am not fully sure of the advantage of providing both ways to do it.

"both" being

1. `EA.factorize`
2. `EA._values_for_factorize`, and `EA._from_factorized`

correct?

Let me see if I can make `mask` as an argument to `factorize` work. That would indeed be cleaner.

jorisvandenbossche commented:
> "both" being
>
>   EA.factorize
>   EA._values_for_factorize, and EA._from_factorized

Yes. Looking at the diff now, it seems to me that overwriting `EA._values_for_factorize` and `EA._from_factorized` will actually need as much as custom code as just overwriting `EA.factorize`. So then the argument of providing `_values_for_factorize` to reduce duplication of the actual `factorize` implementation (like we had for `argsort`) does not fully hold in this case.

> Let me see if I can make mask as an argument to factorize work. That would indeed be cleaner.

Then of course we need to be OK with exposing this in the public `factorize` API. But from an extension author point of view this seems rather clean. 
An alternative would be to specify a `na_value` and then calculate the mask, but I am not sure if that has any advantage over a mask.

@jreback what are your thoughts on such a `mask` keyword for `pd.factorize`

TomAugspurger commented:
I'm having some trouble making the `mask` approach work. It seems like pandas' factorization methods do the right thing for some missing values, like `Decimal(NaN)`.


```python
In [2]: from decimal import Decimal

In [3]: pd.factorize([Decimal('0.0'), Decimal('NaN'), Decimal('0.0')])
Out[3]: (array([ 0, -1,  0]), array([Decimal('0.0')], dtype=object))
```

So in that situation we would have to choose to ignore the mask argument (side-note, `pd.isna` does not consider `Decimal('NaN')` to be NA, so detecting this seems fragile).

jorisvandenbossche commented:
But for those object arrays, we can always use `np.nan` or `None` I think if a mask is specified? That will always work, regardless of whether the actual NA value is recognized by `factorize` ?

jorisvandenbossche commented:
Ah, yes, I was thinking of the "change `values` before passing to `factorize`". 
The problem with that is of course that we then should take a copy of `values`, which might be an unnecessary copy in case the `values` were constructed specifically for `factorize`

TomAugspurger commented:
Here are the cases I see. I put quotes around the ExtensionArray's definition of missing, which may differ from pandas'.

1. float/object values with "NA" values as NA (Decimal)
2. float/object values with "NA" values as something else (IPArray)
3. integer values with "NA" values as iNaT (DatetimeTZ)
4. integer values with "NA" values as something else (Categorical)

You could add a 5/6 for our uint64 hashtable. I don't think that changes anything.

So for cases where the NA values align (1, 3), things will just work, no `mask` required.

For cases where they don't align (2, 4), we'll need to either modify the values going into the hashtable, so that `values[mask] = pandas_na_for_hashtable`, or modify that labels / uniques coming out.

TomAugspurger commented:
Though it occurs to me that a mask with NA values goes against https://github.com/pandas-dev/pandas/issues/20328 a bit... Implementing https://github.com/pandas-dev/pandas/issues/20328 would also make this cleaner, as the EX author need only pass through an `na_value`, and everything should be OK. I'll see if doing that is simple.

jorisvandenbossche commented:
@TomAugspurger Alternative is to just document that if you have float/object data, we expect np.nan, if you have integer data, we expect iNaT as missing value. And that extension authors then apply the mask themselves. 
That brings us back to the original implementation of the PR I think ..

TomAugspurger committed:
Mask values going into hashtables

1. Offer _values_for_factorize
2. Hardcode the NA type going into the hashtable.

TomAugspurger commented:
#20328 doesn't look trivial, so I've abandoned it for now.

> Alternative is to just document that if you have float/object data, we expect np.nan, if you have integer data, we expect iNaT as missing value.

Yes, we'll need to do this regardless, as extension-array authors may have to take precautions if they use iNaT (int min) as a valid value. But I think we'll have to make it clear that this is an implementation detail at the moment.

>  That brings us back to the original implementation of the PR I think ..

Close. The main difference is that in c288d67, we mask the values going *into* the hashtable.

If this approach looks OK, I will add sufficiently scary warnings that the default implementation is using features private to pandas (`_factorize_array`, the missing value markers for each type of hashtable, etc.).

Tangent, I noticed that the UInt64Hashtable doesn't have a missing value marker. Was that intentional (ping @gfyoung I think).

```python
In [9]: from pandas.core.algorithms import _factorize_array

In [10]: _factorize_array(arr, check_nulls=True)
Out[10]:
(array([0, 0, 1, 2]),
 array([                   0,                    1, 18446744073709551615],
       dtype=uint64))
```

In hashtable_helper, the null condition seems to be the string `'False'`.

```
# name, dtype, null_condition, float_group
dtypes = [('Float64', 'float64', 'val != val', True),
          ('UInt64', 'uint64', 'False', False),
          ('Int64', 'int64', 'val == iNaT', False)]
```

Anyway, changing this probably isn't desired, but thought I'd mention it.

gfyoung commented:
> Tangent, I noticed that the UInt64Hashtable doesn't have a missing value marker.

@TomAugspurger : Unfortunately, there is no well-defined missing value marker for `uint` AFAIK.

TomAugspurger committed:
ENH: Parametrized NA sentinel for factorize

Adds a new keyword `na_value` to control the NA sentinel inside the factorize
routine.

```python
In [3]: arr = np.array([0, 1, 0, 2], dtype='u8')

In [4]: pd.factorize(arr)
Out[4]: (array([0, 1, 0, 2]), array([0, 1, 2], dtype=uint64))

In [5]: pd.factorize(arr, na_value=0)
Out[5]: (array([-1,  0, -1,  1]), array([1, 2], dtype=uint64))
```

TomAugspurger committed:
remove stale comment

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -353,6 +371,66 @@ def unique(self):
+        elif com.is_float_dtype(arr) or com.is_object_dtype(arr):
+            arr[mask] = np.nan
+
+        labels, uniques = _factorize_array(arr, check_nulls=True,
```
we could also use `pd.factorize` here if we want to limit the use of internal APIs

TomAugspurger replied:
I was just trying that. Unfortunately, I have to pass `check_nulls=True` for categorical to work properly. It's typically False for integers. But perhaps that's a sign we're doing something improper :/

jorisvandenbossche replied:
Ah, yes, because there we still know what the original dtype was ... 
Hmm, that maybe shows that we rather need a `na_value` keyword (just from a perspective of making `pd.factorize` generally usable, not necessarily in this PR). 
In general, the default of `na_value=None` could be the `val != val` idiom, and you can override that with `val == na_value` with this keyword (e.g. to specify iNaT or -1 for integer data).

TomAugspurger replied:
OK, this issue and the fact that UInt64Hashtable doesn't have a null value are making me come back around to the mask approach.

That seems like the only way we can ensure that we'll actually get things correct.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -353,6 +371,66 @@ def unique(self):
+        if com.is_signed_integer_dtype(arr):
+            arr[mask] = iNaT
+        elif com.is_float_dtype(arr) or com.is_object_dtype(arr):
+            arr[mask] = np.nan
```
My original idea with a `mask` keyword for `factorize`, is that this if/elif code block would do this in `factorize` itself

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -353,6 +371,66 @@ def unique(self):
+            and be a supported dtype. By default, the extension array is cast
+            to object dtype.
+        """
+        return self.astype(object)
```
We should add a comment here that those values might be modified (in the current state of the PR). So that you eg don't just return the ordinals / codes for Period (which is not the case for Categorical right now, because it does `astype(int64)` which always copies)

TomAugspurger committed:
wip

TomAugspurger committed:
REF: Moved to get_labels

TomAugspurger committed:
Remove python-level use_na_value

TomAugspurger committed:
REF: More cleanup

Removed null_condition from the template.

Had Categorical use na_value

TomAugspurger committed:
API: Make it non-public

TomAugspurger committed:
Revert formatting changes in pxd

TomAugspurger commented:
@jreback you can ignore all the back and forth between Joris and I yesterday about `mask`. I think https://github.com/pandas-dev/pandas/pull/20473 will make things nice for EA authors. They'll just need to specify

- The array to be used for factorization
- The scalar in their array to be considered NA.

- defualt: `self.astype(object), np.nan`
- Categorical: `self.codes.astype(int64), -1`
- MACArray: `self.data, 0` (this is in cyberpandas. A uint64 array, where 0 is NA)

TomAugspurger committed:
linting

TomAugspurger committed:
Handle bool

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into parametrized-factorize-na-value

TomAugspurger committed:
Specify dtypes

TomAugspurger committed:
Remove unused variable.

Added PyObject hashtable test

TomAugspurger committed:
REF: Removed check_nulls

TomAugspurger committed:
BUG: NaT for period

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into parametrized-factorize-na-value

TomAugspurger committed:
Other hashtable

TomAugspurger committed:
na_value_for_dtype PeriodDtype

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger committed:
Merge branch 'parametrized-factorize-na-value' into ea-factorize-2+parametrized

TomAugspurger committed:
Use na_value

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-factorize-2

TomAugspurger committed:
Typing

Spacing

Docs

jorisvandenbossche commented:
> @jorisvandenbossche do you think this will work well for geopandas? Can you have missing geometries?

Yes, that should be fine. I am not fully sure yet, but I think I will use a binary blob for each geometry ("well known binary" format), which will give an object array of bytes, and with eg None as missing value. So this should work (that should already work with `pd.factorize` itself without the new infrastructure I think). 
I will also need the `original` to add back metadata about the coordinate reference system.

TomAugspurger commented:
From Python your array of pointers look like Int64s, right? Could those be factorized? I *think* it should, though I haven't thought it through fully.

jorisvandenbossche commented:
No, because two separate Points (separate objects) but with same coordinates, will have different pointer. So the pointer information is not necessarily relevant for factorize.

TomAugspurger commented:
Understood.

CI is all green.

jreback commented:
thanks @TomAugspurger
